contracts/src/spaces/facets/membership/MembershipFacet.sol

```diff
@@ -147,15 +150,32 @@ contract MembershipFacet is
       (address, address)
     );
 
-    // allocate protocol and membership fees
-    uint256 membershipPrice = _getMembershipPrice(_totalSupply());
+    uint256 totalSupply = _totalSupply();
+    uint256 membershipPrice;
+
+    uint256 freeAllocation = _getMembershipFreeAllocation();
+    uint256 prepaidSupply = _getPrepaidSupply();
+
+    if (freeAllocation > totalSupply) {
+      membershipPrice = 0;
+      _refundBalance(transactionId, sender);
+    } else if (prepaidSupply > 0) {
+      membershipPrice = 0;
+      _reducePrepay(1);
+      _refundBalance(transactionId, sender);
+    } else {
+      membershipPrice = _getMembershipPrice(totalSupply);
+    }
+
+    // get token id
     uint256 tokenId = _nextTokenId();
 
     if (membershipPrice > 0) {
       uint256 userValue = _getCapturedValue(transactionId);
 
       if (userValue == 0) revert Membership__InsufficientPayment();
-      if (membershipPrice > userValue) revert Membership__InsufficientPayment();
+      if (userValue != membershipPrice)
+        revert Membership__InsufficientPayment();
```

Comment:
This error is misleading when `userValue > membershipPrice`. Could we have a new error `Membership__OverPayment()` to distinguish between the two cases more clearly or refund `userValue - membershipPrice` at the end if `userValue > membershipPrice` ?